File: tests/corpus/concept.cs
// vyakarana stand-in -- not from vidya. See docs/adr/0006-standin-corpus-policy.md.
// Re-sync when vidya adds a C# reference sample.
//
// Lexing + parsing of arithmetic expressions, modern C# idioms.

using System;
using System.Collections.Generic;

namespace Vyakarana.ConceptParser;

// ── Tokens ───────────────────────────────────────────────────────────

public enum TokenKind
{
    Number,
    Plus,
    Minus,
    Star,
    Slash,
    LParen,
    RParen,
    Eof,
}

public readonly record struct Token(TokenKind Kind, string Text, int Pos);

// ── Lexer ────────────────────────────────────────────────────────────

public sealed class Lexer
{
    private readonly string _source;
    private int _pos;

    public Lexer(string source)
    {
        _source = source;
        _pos = 0;
    }

    private void SkipWhitespace()
    {
        while (_pos < _source.Length && char.IsWhiteSpace(_source[_pos]))
        {
            _pos++;
        }
    }

    public Token Next()
    {
        SkipWhitespace();
        var start = _pos;

        if (_pos >= _source.Length)
        {
            return new Token(TokenKind.Eof, "", _pos);
        }

        var c = _source[_pos];
        switch (c)
        {
            case '+': _pos++; return new Token(TokenKind.Plus,   "+", start);
            case '-': _pos++; return new Token(TokenKind.Minus,  "-", start);
            case '*': _pos++; return new Token(TokenKind.Star,   "*", start);
            case '/': _pos++; return new Token(TokenKind.Slash,  "/", start);
            case '(': _pos++; return new Token(TokenKind.LParen, "(", start);
            case ')': _pos++; return new Token(TokenKind.RParen, ")", start);
        }

        if (char.IsDigit(c))
        {
            while (_pos < _source.Length && char.IsDigit(_source[_pos]))
            {
                _pos++;
            }
            return new Token(TokenKind.Number, _source[start.._pos], start);
        }

        throw new InvalidOperationException($"unexpected character at {_pos}");
    }
}

// ── AST ──────────────────────────────────────────────────────────────

public abstract record Expr
{
    public abstract long Eval();
}

public sealed record NumberExpr(long Value) : Expr
{
    public override long Eval() => Value;
}

public sealed record UnaryExpr(char Op, Expr Operand) : Expr
{
    public override long Eval() => -Operand.Eval();
}

public sealed record BinOpExpr(char Op, Expr Left, Expr Right) : Expr
{
    public override long Eval() => Op switch
    {
        '+' => Left.Eval() + Right.Eval(),
        '-' => Left.Eval() - Right.Eval(),
        '*' => Left.Eval() * Right.Eval(),
        '/' => Left.Eval() / Right.Eval(),
        _   => throw new InvalidOperationException($"bad op {Op}"),
    };
}

// ── Pratt parser ─────────────────────────────────────────────────────

public sealed class Parser
{
    private readonly Lexer _lexer;
    private Token _current;

    public Parser(string source)
    {
        _lexer = new Lexer(source);
        _current = _lexer.Next();
    }

    private void Advance() { _current = _lexer.Next(); }

    private static int InfixBp(TokenKind k) => k switch
    {
        TokenKind.Plus or TokenKind.Minus => 1,
        TokenKind.Star or TokenKind.Slash => 3,
        _                                  => 0,
    };

    public Expr ParseExpr(int minBp)
    {
        var left = ParsePrimary();
        while (true)
        {
            var bp = InfixBp(_current.Kind);
            if (bp == 0 || bp < minBp) break;
            var op = _current.Text[0];
            Advance();
            var right = ParseExpr(bp + 1);
            left = new BinOpExpr(op, left, right);
        }
        return left;
    }

    private Expr ParsePrimary()
    {
        if (_current.Kind == TokenKind.Minus)
        {
            Advance();
            return new UnaryExpr('-', ParsePrimary());
        }
        if (_current.Kind == TokenKind.LParen)
        {
            Advance();
            var inner = ParseExpr(0);
            if (_current.Kind != TokenKind.RParen)
            {
                throw new InvalidOperationException("expected )");
            }
            Advance();
            return inner;
        }
        if (_current.Kind == TokenKind.Number)
        {
            var v = long.Parse(_current.Text);
            Advance();
            return new NumberExpr(v);
        }
        throw new InvalidOperationException($"unexpected token {_current}");
    }
}

// ── Demo ─────────────────────────────────────────────────────────────

public static class Program
{
    public static void Main()
    {
        var cases = new List<(string Input, long Expected)>
        {
            ("1 + 2",       3L),
            ("2 * 3 + 4",   10L),
            ("(1 + 2) * 3", 9L),
            ("-5 + 3",      -2L),
        };
        foreach (var (input, expected) in cases)
        {
            var got = new Parser(input).ParseExpr(0).Eval();
            Console.WriteLine($"{input,-15} => {got} (expected {expected})");
            if (got != expected)
            {
                throw new InvalidOperationException($"mismatch at \"{input}\"");
            }
        }
    }
}

// 2.3.2 error-hole coverage: C# preprocessor directives.
#nullable enable
#region Diagnostics
#if DEBUG
#pragma warning disable CS0168
#endif
#endregion

File: tests/corpus/concept.kt
// vyakarana stand-in -- not from vidya. See docs/adr/0006-standin-corpus-policy.md.
// Re-sync when vidya adds a Kotlin reference sample.
//
// Lexing + parsing of arithmetic expressions, Kotlin idioms.

package com.example.vyakarana

import kotlin.text.StringBuilder

// ── Tokens ───────────────────────────────────────────────────────────

enum class TokenKind {
    NUMBER, PLUS, MINUS, STAR, SLASH, LPAREN, RPAREN, EOF
}

data class Token(val kind: TokenKind, val text: String, val pos: Int) {
    override fun toString(): String = "Token($kind, $text, $pos)"
}

// ── Lexer ────────────────────────────────────────────────────────────

class Lexer(private val source: String) {
    private var pos: Int = 0

    private fun peek(): Char =
        if (pos < source.length) source[pos] else '_'

    private fun skipWhitespace() {
        while (pos < source.length && source[pos].isWhitespace()) {
            pos++
        }
    }

    fun next(): Token {
        skipWhitespace()
        val start = pos

        if (pos >= source.length) {
            return Token(TokenKind.EOF, "", pos)
        }

        val c = source[pos]
        when (c) {
            '+' -> { pos++; return Token(TokenKind.PLUS,   "+", start) }
            '-' -> { pos++; return Token(TokenKind.MINUS,  "-", start) }
            '*' -> { pos++; return Token(TokenKind.STAR,   "*", start) }
            '/' -> { pos++; return Token(TokenKind.SLASH,  "/", start) }
            '(' -> { pos++; return Token(TokenKind.LPAREN, "(", start) }
            ')' -> { pos++; return Token(TokenKind.RPAREN, ")", start) }
        }

        if (c.isDigit()) {
            while (pos < source.length && source[pos].isDigit()) {
                pos++
            }
            return Token(TokenKind.NUMBER, source.substring(start, pos), start)
        }

        error("unexpected character '$c' at $pos")
    }
}

// ── AST ──────────────────────────────────────────────────────────────

sealed interface Expr
data class NumberExpr(val value: Long) : Expr
data class BinOpExpr(val op: Char, val left: Expr, val right: Expr) : Expr
data class UnaryExpr(val op: Char, val operand: Expr) : Expr

// ── Pratt parser ─────────────────────────────────────────────────────

class Parser(source: String) {
    private val lexer = Lexer(source)
    private var current: Token = lexer.next()

    private fun advance(): Token {
        val prev = current
        current = lexer.next()
        return prev
    }

    private fun infixBp(k: TokenKind): Int = when (k) {
        TokenKind.PLUS, TokenKind.MINUS -> 1
        TokenKind.STAR, TokenKind.SLASH -> 3
        else -> 0
    }

    fun parseExpr(minBp: Int): Expr {
        var left: Expr = parsePrimary()
        while (true) {
            val bp = infixBp(current.kind)
            if (bp == 0 || bp < minBp) break
            val op = current.text[0]
            advance()
            val right = parseExpr(bp + 1)
            left = BinOpExpr(op, left, right)
        }
        return left
    }

    private fun parsePrimary(): Expr {
        if (current.kind == TokenKind.MINUS) {
            advance()
            return UnaryExpr('-', parsePrimary())
        }
        if (current.kind == TokenKind.LPAREN) {
            advance()
            val inner = parseExpr(0)
            check(current.kind == TokenKind.RPAREN) { "expected )" }
            advance()
            return inner
        }
        if (current.kind == TokenKind.NUMBER) {
            val v = current.text.toLong()
            advance()
            return NumberExpr(v)
        }
        error("unexpected token $current")
    }
}

// ── Evaluator ────────────────────────────────────────────────────────

fun eval(e: Expr): Long = when (e) {
    is NumberExpr -> e.value
    is UnaryExpr  -> -eval(e.operand)
    is BinOpExpr  -> when (e.op) {
        '+' -> eval(e.left) + eval(e.right)
        '-' -> eval(e.left) - eval(e.right)
        '*' -> eval(e.left) * eval(e.right)
        '/' -> eval(e.left) / eval(e.right)
        else -> error("op ${e.op}")
    }
}

// ── Demo ─────────────────────────────────────────────────────────────

fun main() {
    val cases = listOf(
        "1 + 2"          to 3L,
        "2 * 3 + 4"      to 10L,
        "(1 + 2) * 3"    to 9L,
        "-5 + 3"         to -2L,
    )
    for ((input, expected) in cases) {
        val got = eval(Parser(input).parseExpr(0))
        println("%-15s => %d (expected %d)".format(input, got, expected))
        require(got == expected) { "mismatch at \"$input\"" }
    }
}

// 2.3.2 error-hole coverage: backtick-escaped identifiers. The
// test-naming idiom below is the most common Kotlin use of them.
fun `tokenizes an empty input`() { }

val `object` = "reserved word used as a name"
